linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala

```diff
@@ -1198,17 +1252,50 @@ final class CoreWasmLib(coreSpec: CoreSpec) {
 
     fb.block() { objIsNullLabel =>
       primType match {
+        // For byte and short, use br_on_cast_fail with i31 then check the value
+        case ByteType | ShortType =>
+          val intValueLocal = fb.addLocal("intValue", Int32)
+
+          fb.block(RefType.anyref) { castFailLabel =>
+            fb += LocalGet(objParam)
+            fb += BrOnNull(objIsNullLabel)
+            fb += BrOnCastFail(castFailLabel, RefType.any, RefType.i31)
+
+            // Extract the i31 value
+            fb += I31GetS
+            fb += LocalTee(intValueLocal)
+
+            // if it sign-extends to itself
+            fb += LocalGet(intValueLocal)
+            if (primType == ByteType)
+              fb += I32Extend8S
+            else
+              fb += I32Extend16S
+            fb += I32Eq
+            fb.ifThen() {
+              // then success
+              if (isUnbox)
+                fb += LocalGet(intValueLocal)
+              else
+                fb += LocalGet(objParam)
+              fb += Return
+            }
+
+            // Fall through for CCE
+            fb += LocalGet(objParam)
+          }
+
         // For char and long, use br_on_cast_fail to test+cast to the box class
         case CharType | LongType =>
           val boxClass =
             if (primType == CharType) SpecialNames.CharBoxClass
             else SpecialNames.LongBoxClass
           val structTypeID = genTypeID.forClass(boxClass)
 
-          fb.block(RefType.anyref) { castFailLabel =>
+          fb.block(RefType.any) { castFailLabel =>
             fb += LocalGet(objParam)
             fb += BrOnNull(objIsNullLabel)
-            fb += BrOnCastFail(castFailLabel, RefType.anyref, RefType(structTypeID))
+            fb += BrOnCastFail(castFailLabel, RefType.any, RefType(structTypeID))
```

Comment:
Wrong commit? (feels like this was better all along?)

Reply:
Good point. I extracted in a separate commit.